File: XRViewer/Animator/Animator.m
#import "Animator.h"
#import <pop/POP.h>

#define ANIMATION_PULSE_KEY @"pulse"
#define ANIMATION_FRAME_KEY @"frame"
#define ANIMATION_COLOR_KEY @"color"

@interface AnimationDelegate : NSObject<CAAnimationDelegate>
@property(nonatomic, copy) Completion completion;
@end

@interface Animator () <CAAnimationDelegate>
@property(nonatomic, strong) NSMutableArray *animationCompletions;
@end

@implementation Animator

- (void)dealloc
{
    DDLogDebug(@"Animator dealloc");
}

- (instancetype)init
{
    self = [super init];
    
    if (self)
    {
        [self setAnimationCompletions:[NSMutableArray new]];
        [self setAnimationDuration:DEFAULT_ANIMATION_DURATION];
    }
    
    return self;
}

- (void)clean
{
    [[self animationCompletions] removeAllObjects];
    [[[UIApplication sharedApplication] keyWindow] pop_removeAllAnimations];
    [[[[UIApplication sharedApplication] keyWindow] layer] pop_removeAllAnimations];
    [[[[UIApplication sharedApplication] keyWindow] layer] removeAllAnimations];
}

- (void)startPulseAnimation:(UIView *)view
{
    POPSpringAnimation *anim = [POPSpringAnimation animationWithPropertyNamed:kPOPViewScaleXY];
    anim.toValue = @(CGPointMake(1.1, 1.1));
    anim.fromValue = @(CGPointMake(0.9 , 0.9));
    anim.repeatForever = YES;
    anim.autoreverses = YES;
    
    [view pop_addAnimation:anim forKey:ANIMATION_PULSE_KEY];
}

- (void)stopPulseAnimation:(UIView *)view
{
    [view pop_removeAnimationForKey:ANIMATION_PULSE_KEY];
}

- (void)animate:(UIView *)view toFrame:(CGRect)frame
{
    [self animate:view toFrame:frame completion:NULL];
}

- (void)animate:(UIView *)view toFrame:(CGRect)frame completion:(Completion)completion
{
    if (CGRectEqualToRect(frame, [view frame]))
    {
        if (completion) {
            dispatch_async(dispatch_get_main_queue(), ^
                           {
                               completion(NO);
                           });
        }
        
        return;
    }
    
    POPSpringAnimation *anim = [POPSpringAnimation animationWithPropertyNamed:kPOPViewFrame];
    anim.toValue = @(frame);
    anim.fromValue = @([view frame]);
    [anim setCompletionBlock:^(POPAnimation *anim, BOOL finished)
     {
         [view setFrame:frame];
         
         if (completion)
         {
             completion(finished);
         }
     }];
    
    [view pop_addAnimation:anim forKey:ANIMATION_FRAME_KEY];
}

- (void)animate:(UIView *)view toFade:(BOOL)fade
{
    [self animate:view toFade:fade completion:NULL];
}

- (void)animate:(UIView *)view toFade:(BOOL)fade completion:(Completion)completion
{
    CGFloat newOpacity = fade? 0 : 1;
    
    if ([[view layer] opacity] == newOpacity)
    {
        if (completion) {
            dispatch_async(dispatch_get_main_queue(), ^
                           {
                               completion(NO);
                           });
        }
        
        return;
    }
    
    [[view layer] setOpacity:newOpacity];
    
    NSString *key = [NSString stringWithFormat:@"FADE-%@", view];
    [[view layer] removeAnimationForKey:key];
    
    CATransition *transition = [CATransition animation];
    [transition setDuration:[self animationDuration]];
    [transition setType:kCATransitionFade];
    
    AnimationDelegate *ad = [AnimationDelegate new];
    __weak AnimationDelegate *blockAd = ad;
    __weak Animator *blockSelf = self;
    
    [ad setCompletion:^(BOOL f)
     {
         if (completion)
         {
             completion(f);
         }
         
         [[blockSelf animationCompletions] removeObject:blockAd];
     }];
    
    [transition setDelegate:ad];
    [[self animationCompletions] addObject:ad];
    
    [[view layer] addAnimation:transition forKey:key];
}

- (void)animate:(UIView *)view toColor:(UIColor *)color
{
    if ( [view backgroundColor] == color )
    {
        return;
    }
    
    POPSpringAnimation *anim = [POPSpringAnimation animationWithPropertyNamed:kPOPViewBackgroundColor];
    anim.toValue = color;
    anim.fromValue = [view backgroundColor];
    [anim setCompletionBlock:^(POPAnimation *anim, BOOL finished)
     {
         [view setBackgroundColor:color];
     }];
    
    [view pop_addAnimation:anim forKey:ANIMATION_COLOR_KEY];
}

@end



@implementation AnimationDelegate
- (void)animationDidStop:(CAAnimation *)anim finished:(BOOL)flag
{
    if ([self completion])
    {
        [self completion](flag);
    }
}
@end

File: XRViewer/Utilities/Animator.swift
import UIKit
import XCGLogger

typealias Completion = (Bool) -> Void

let DEFAULT_ANIMATION_DURATION = 0.5
let ANIMATION_PULSE_KEY = "pulse"
let ANIMATION_FRAME_KEY = "frame"
let ANIMATION_COLOR_KEY = "color"

class AnimationDelegate: NSObject, CAAnimationDelegate {
    var completion: Completion?
    
    func animationDidStop(_ anim: CAAnimation, finished flag: Bool) {
        //if completion
        
        completion?(flag)
    }
}

class Animator: NSObject, CAAnimationDelegate {
    
    var animationDuration: Double = 0.0
    private var animationCompletions: [AnimationDelegate] = []
    
    override init() {
        super.init()
        
        self.animationCompletions = []
        animationDuration = DEFAULT_ANIMATION_DURATION
    }
    
    deinit {
        appDelegate().logger.debug("Animator dealloc")
    }
    
    func clean() {
        animationCompletions.removeAll()
        UIApplication.shared.keyWindow?.layer.removeAllAnimations()
    }

    func animate(_ view: UIView?, toFade fade: Bool) {
        animate(view, toFade: fade) { (bool) in
        }
    }

    func animate(_ view: UIView?, toFade fade: Bool, completion: @escaping Completion) {
        let newOpacity: CGFloat = fade ? 0 : 1

        if CGFloat(view?.layer.opacity ?? 0.0) == newOpacity {
            //if completion

            DispatchQueue.main.async(execute: {
                completion(false)
            })
            return
        }

        view?.layer.opacity = Float(newOpacity)

        var key: String? = nil
        if let aView = view {
            key = "FADE-\(aView)"
        }
        view?.layer.removeAnimation(forKey: key ?? "")

        let transition = CATransition()
        transition.duration = animationDuration
        transition.type = .fade

        let ad = AnimationDelegate()
        weak var blockAd: AnimationDelegate? = ad
        weak var blockSelf: Animator? = self

        ad.completion = { f in
            //if completion

            completion(f)
            blockSelf?.animationCompletions.removeAll(where: { $0 == blockAd })
        }

        transition.delegate = ad
        animationCompletions.append(ad)

        view?.layer.add(transition, forKey: key)
    }
}
